libraries/aws/provisioning/test/unit/aws_iot_tests_provisioning_serializer.c

```diff
@@ -64,6 +96,59 @@ static const size_t _numOfSampleParameters = 3;
 
 /*-----------------------------------------------------------*/
 
+/**
+ * @brief Allocates memory for buffer used in tests for serialization
+ * with reserve space to check against buffer overrun errors by serializer
+ * functions under test.
+ * @note The reserve space is used at the beginning and the end of the
+ * allocated buffer space.
+ */
+uint8_t * _allocateBufferMemoryWithBufferOverrunReserve( size_t sizeOfSerialization,
+                                                         size_t * serializationStartingIndex )
+{
+    /* We will allocate more space than the expected serialization size to perform
+     * buffer overrun checks after the serialization operation.  */
+    size_t totalBufferSize = sizeOfSerialization +
+                             _reserveSize + _reserveSize;
+    uint8_t * testBuffer = unity_malloc_mt( totalBufferSize );
+
+    assert( testBuffer != NULL );
+
+    /* Fill the buffer with the buffer overrun test value which will be */
+    /* checked after the serialization call. */
+    memset( testBuffer, _bufferOverrunCheckValue, totalBufferSize );
+
+    *serializationStartingIndex = _reserveSize;
+
+    return testBuffer;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Verifies that the reserve space in test buffer used for serialization
+ * is not corrupted by buffer overrun errors in the serializer function under
+ * test.
+ */
+void _checkForBufferOverrun( const uint8_t * buffer,
+                             size_t sizeOfSerialization )
+{
+    size_t totalBufferSize = sizeOfSerialization +
+                             _reserveSize + _reserveSize;
+
+    for( int index = 0; index < _reserveSize; index++ )
+    {
+        assert( _bufferOverrunCheckValue == buffer[ index ] );
+    }
+
+    for( int index = totalBufferSize - 1; index >= totalBufferSize - _reserveSize; index-- )
+    {
+        assert( _bufferOverrunCheckValue == buffer[ index ] );
+    }
```

Comment:
Looks like these two `for`s could be combined into one. Just a style choice though, and it doesn't really matter since this is only a unit test file.

Reply:
Had done for better readability. I suppose that having in the same loop would be readable as well with comments. Made the change